web/src/copy_and_paste.js

```diff
@@ -293,29 +293,138 @@ function get_end_tr_from_endc($endc) {
+    // turning off escaping (for now) to remove extra `/`
+    TurndownService.prototype.escape = (string) => string;
+
+    const turndownService = new TurndownService({emDelimiter: "*", codeBlockStyle: "fenced"});
     turndownService.addRule("headings", {
         filter: ["h1", "h2", "h3", "h4", "h5", "h6"],
         replacement(content) {
             return content;
         },
     });
-    turndownService.addRule("emphasis", {
-        filter: ["em", "i"],
+    turndownService.addRule("strikethrough", {
+        filter: ["del", "s", "strike"],
         replacement(content) {
-            return "*" + content + "*";
+            return "~~" + content + "~~";
         },
     });
-    // Checks for raw links without custom text or title.
     turndownService.addRule("links", {
+        filter: ["a"],
+        replacement(content, node) {
+            if (node.href === node.innerHTML) {
+                // Checks for raw links without custom text.
+                return content;
+            }
+            if (node.childNodes.length === 1 && node.firstChild.nodeName === "IMG") {
+                // ignore link's url if it only has an image
+                return content;
+            }
+            return "[" + content + "](" + node.href + ")";
+        },
+    });
+    turndownService.addRule("listItem", {
+        filter: "li",
+        replacement(content, node) {
+            content = content
+                .replace(/^\n+/, "") // remove leading newlines
+                .replace(/\n+$/, "\n") // replace trailing newlines with just a single one
+                .replace(/\n/gm, "\n  "); // custom 2 space indent
+            let prefix = "* ";
+            const parent = node.parentNode;
+            if (parent.nodeName === "OL") {
+                const start = parent.getAttribute("start");
+                const index = Array.prototype.indexOf.call(parent.children, node);
+                prefix = (start ? Number(start) + index : index + 1) + ". ";
+            }
+            return prefix + content + (node.nextSibling && !/\n$/.test(content) ? "\n" : "");
+        },
+    });
+    turndownService.addRule("zulipCodeBlock", {
+        filter(node, options) {
+            return (
+                options.codeBlockStyle === "fenced" &&
+                node.nodeName === "CODE" &&
+                node.parentElement?.nodeName === "PRE"
+            );
+        },
+
+        replacement(content, node, options) {
+            const language = node.closest(".codehilite")?.dataset?.codeLanguage || "";
+
+            const fenceChar = options.fence.charAt(0);
+            let fenceSize = 3;
+            const fenceInCodeRegex = new RegExp("^" + fenceChar + "{3,}", "gm");
+
+            let match;
+            while ((match = fenceInCodeRegex.exec(content))) {
+                if (match[0].length >= fenceSize) {
+                    fenceSize = match[0].length + 1;
+                }
+            }
```

Comment:
This fencing logic is likely to not work well with nested code blocks. Does the built-in Turndown logic behave better by any chance?

Hmm, skimming their code, this logic appears to be borrowed from upstream. It's important for licensing and understanding reasons to document this sort of detail -- can you write a comment noting what you've changed from the upstream implementation?

Is it just the calculation of `var language` or something more?

Reply:
Hmm, OK, that's probably a reasonable filter if it's easy to write correctly, but in any case please write down that reasoning in a comment; it'd be pretty hard to figure out without your having said it that was what was going on.